File: dotnet/samples/01-get-started/03_MultiTurn.cs
// Copyright (c) Microsoft. All rights reserved.

// Step 3: Multi-Turn Conversations
// Maintain conversation context across multiple exchanges with the agent.
// The session object preserves conversation state between calls.
//
// For more on conversations, see: ../02-agents/conversations/
// For docs: https://learn.microsoft.com/agent-framework/agents/conversations

using Azure.AI.Projects;
using Azure.AI.Projects.OpenAI;
using Azure.Identity;
using Microsoft.Agents.AI;

string endpoint = Environment.GetEnvironmentVariable("AZURE_FOUNDRY_PROJECT_ENDPOINT")
    ?? throw new InvalidOperationException("AZURE_FOUNDRY_PROJECT_ENDPOINT is not set.");
string deploymentName = Environment.GetEnvironmentVariable("AZURE_FOUNDRY_PROJECT_DEPLOYMENT_NAME") ?? "gpt-4o-mini";

AIProjectClient aiProjectClient = new(new Uri(endpoint), new AzureCliCredential());

ChatClientAgent agent = await aiProjectClient.CreateAIAgentAsync(
    name: "JokerAgent",
    options: new AgentVersionCreationOptions(
        new PromptAgentDefinition(model: deploymentName)
        {
            Instructions = "You are good at telling jokes."
        }));

// <multi_turn>
// Create a conversation for tracking in Foundry UI
ProjectConversationsClient conversationsClient = aiProjectClient.GetProjectOpenAIClient().GetProjectConversationsClient();
ProjectConversation conversation = await conversationsClient.CreateProjectConversationAsync();

AgentSession session = await agent.CreateSessionAsync(conversation.Id);

Console.WriteLine(await agent.RunAsync("Tell me a joke about a pirate.", session));
Console.WriteLine(await agent.RunAsync("Now add some emojis to the joke and tell it in the voice of a pirate's parrot.", session));
// </multi_turn>

// Cleanup
await aiProjectClient.Agents.DeleteAgentAsync(agent.Name);
await conversationsClient.DeleteConversationAsync(conversation.Id);

File: python/samples/01-get-started/03_multi_turn.py
# Copyright (c) Microsoft. All rights reserved.

import asyncio
import os

from agent_framework.openai import OpenAIResponsesClient

"""
Multi-Turn Conversations — Use AgentThread to maintain context

This sample shows how to keep conversation history across multiple calls
by reusing the same thread object.
"""


async def main() -> None:
    # <create_agent>
    client = OpenAIResponsesClient(
        api_key=os.environ["OPENAI_API_KEY"],
        model_id=os.environ.get("OPENAI_RESPONSES_MODEL_ID", "gpt-4o"),
    )

    agent = client.as_agent(
        name="ConversationAgent",
        instructions="You are a friendly assistant. Keep your answers brief.",
    )
    # </create_agent>

    # <multi_turn>
    # Create a thread to maintain conversation history
    thread = agent.get_new_thread()

    # First turn
    result = await agent.run("My name is Alice and I love hiking.", thread=thread)
    print(f"Agent: {result}\n")

    # Second turn — the agent should remember the user's name and hobby
    result = await agent.run("What do you remember about me?", thread=thread)
    print(f"Agent: {result}")
    # </multi_turn>


if __name__ == "__main__":
    asyncio.run(main())
